### DISCUSSION TITLE
Client side SSE with Edge Function Streams

### DISCUSSION THREAD

**Discussion by gregnr** [supabase/supabase#13124] [Questions] [answered]
Carrying over conversation from https://github.com/supabase/supabase/pull/12499 here to discuss how to consume SSE from edge function streams.

## Original question
How can you integrate with Edge Function streams (server-sent events) from the client side?

**Comment by JSmithOner** [comment]
@gregnr Thanks for your reply you are correct `sse.js` does not throw errors on import I was installing the wrong package.
Now my problem is that even after adding an event listener to all kind of message my console prints nothing and my network tab shows no external connections. 

**EDIT**
I needed to add `source.stream()`to trigger streaming.
But now I get cors policy erros event though I've setted fetch on server side as so

```
  return ( fetch("https://api.openai.com/v1/chat/completions", {

    headers: {

      Authorization:`Bearer ${ Deno.env.get("OPEN_AI_KEY") }`,
      "Content-Type":'application/json',
      "Access-Control-Allow-Origin": "*",
      "Access-Control-Allow-Methods": "POST",
      "Access-Control-Expose-Headers": "Content-Length, X-JSON",
      "Access-Control-Allow-Headers": "apikey,X-Client-Info, Content-Type, Authorization, Accept, Accept-Language, X-Authorization",

    },
    method: "POST",
    body: JSON.stringify( {

      //model:'code-davinci-002',
      model:'gpt-4',
      prompt:text,
      temperature:0,
      max_tokens,
      stream:true

    } )

  } ) );
```

**Reply by gregnr** [reply]
Glad you were able to solve the package issue.

Are you getting the CORS errors on your client side? You will need to make sure your Edge Function itself returns the appropriate CORS headers. Check out the Clippy implementation here:
https://github.com/supabase/supabase/blob/b2625c96d8b64855696f6d2e13518362a4701bd2/supabase/functions/clippy-search/index.ts#L153-L158

**Comment by JSmithOner** [comment]
@gregnr it's on its way to work but now I'm enable to use your code to properly catch error types. `common` folder is not on github and I'm unable to read the error from openai api.any code?I'm really sorry I won't bother when this will be solved :). Best.

**Reply by gregnr** [reply]
The `common` folder is available here:
https://github.com/supabase/supabase/tree/b2625c96d8b64855696f6d2e13518362a4701bd2/supabase/functions/common

Keep in mind you don't have to follow the same pattern. That folder was just to share common logic between edge functions.

**Comment by JSmithOner** [comment]
@gregnr Thnaks anyway I'll test and get back to you.

**Comment by gregnr** [comment, marked as answer]
### Edge function example
https://github.com/supabase/supabase/tree/ab94d9ee5f9cad2343b7905046c1ce981f28743d/examples/edge-functions/supabase/functions/streams

### Clippy frontend implementation
We currently use streams in the Clippy search implementation for Supabase docs. You can reference that code here:
https://github.com/supabase/supabase/blob/b2625c96d8b64855696f6d2e13518362a4701bd2/apps/docs/components/Search/SearchModal.tsx#L116-L123

This uses the [sse.js](https://github.com/mpetazzoni/sse.js) library to connect to the SSE stream. Browsers already have an [EventSource](https://developer.mozilla.org/en-US/docs/Web/API/EventSource) API available, but it is limited to `GET` requests and can't pass custom headers. [sse.js](https://github.com/mpetazzoni/sse.js) adds these missing pieces.

[sse.js](https://github.com/mpetazzoni/sse.js) can be installed using the regular NPM install command:
```bash
npm install sse.js
```
and imported as you would any other package. In a ESM environment, you would do:
```js
import { SSE } from 'sse.js'

const source = new SSE(...)
```

**Comment by Pasta** [comment]
Hi @gregnr , reviving this issue because I'm facing similar issues for a few days now. I've followed the clippy example to the letter.

My preflight query returns 200 but somehow returns empty headers and my actual query returns 404. 

Generic headers in the browser but also when I'm checking logs in supaase.co/my-project/logs/edge-logs.

Here's what i see the logs returning in the edge logs: 
Response Metadata
``` json
{
  "headers": [],
  "origin_time": 107,
  "status_code": 200
}

```
here are my logs on the browser for preflight:
```
access-control-allow-headers: apikey,authorization,content-type
access-control-allow-methods: GET,HEAD,PUT,PATCH,POST,DELETE,OPTIONS,TRACE,CONNECT
access-control-allow-origin: *
access-control-max-age: 3600
alt-svc: h3=":443"; ma=86400, h3-29=":443"; ma=86400
cf-ray: 7c357a208eb5fa72-SJC
content-length: 0
date: Sun, 07 May 2023 00:51:53 GMT
server: cloudflare
strict-transport-security: max-age=2592000; includeSubDomains
vary: Accept-Encoding

```
Here's my code on the server side:


``` typescript
const cors = {
  'Access-Control-Allow-Origin': "*",
  "Access-Control-Expose-Headers": "Content-Length, X-JSON",
  "Access-Control-Allow-Headers": "apikey,X-Client-Info, Content-Type, Authorization, Accept, Accept-Language, X-Authorization",
}
serve(async(req: Request) => {
    if (req.method === 'OPTIONS') {
      return new Response('ok', { headers: cors })
    }
[...]
  try {
    const response = await fetch('https://api.openai.com/v1/completions', {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${Deno.env.get('OPENAI_API_KEY')}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(completionConfig),
      });
      return new Response(response.body, {
        headers: {
          ...cors,
          'Content-Type': 'text/event-stream',
        },
      })
    } catch (error) {
      console.error(error);
      return new Response(error, {
        headers: {
          ...cors,
          'Content-Type': 'text/event-stream',
        },
      })
    }
```

and my client is also pretty generic:

``` typescript

const eventSource = new SSE(streamEndpoint, {
        headers: {
          apikey: process.env.REACT_APP_SUPABASE_API_KEY,
          Authorization: `Bearer ${process.env.REACT_APP_SUPABASE_API_KEY}`,
          'Content-Type': 'application/json',
        },
        payload: JSON.stringify({ query }),
      })
      eventSource.addEventListener('message', (e) => {
        if (e.data === '[DONE]') {
          ZZZZ
          return;
        }
        const data = JSON.parse(e.data);
        setStreamData((prevData) => prevData + data.choices[0].text);
      })
      eventSource.addEventListener('error', (e) => {
        console.log(e);
      }) 
```

It's working locally, but I can't get it to work other wise. What am I forgetting here? 

If I curl the function, it works as expected pointing to a CORS issue and not something else.

**Reply by gregnr** [reply]
Hey @Pasta, what are you using for `streamEndpoint` on the client side? There are some differences in the URL format between local and cloud.

Assuming that's not the issue, might be worth changing your error response type to `application/json` instead of `text/event-stream` - this could be why you aren't able to see any useful error information client side.

**Reply by Pasta** [reply]
Hey @gregnr 
my streamEndpoint looks like https://project-id.supabase.co/my-edge-function-name

my query still returns 404 with this error:
```
Access to XMLHttpRequest at 'https://project-id.supabase.co/my-edge-function-name' from origin 'https://xxxx.vercel.app' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
```

Which is weird looking at my previous code. I also see this in the console.log(e)
```
{
    "isTrusted": false,
    "data": "",
    "source": {
        "INITIALIZING": -1,
        "CONNECTING": 0,
        "OPEN": 1,
        "CLOSED": 2,
        "url": "https://project-id.supabase.co/my-edge-function-name",
        "headers": {
            "apikey": "correct-api-key",
            "Authorization": "Bearer correct-api-key",
            "Content-Type": "application/json"
        },
        "payload": "{payload}",
        "method": "POST",
        "withCredentials": false,
        "FIELD_SEPARATOR": ":",
        "listeners": {
            "message": [
                null
            ],
            "error": [
                null
            ]
        },
        "xhr": {},
        "readyState": 2,
        "progress": 0,
        "chunk": ""
    }
}

**Reply by gregnr** [reply]
URL looks good.

Very bizarre you are getting the 404 & CORS error despite including the headers. Another potential way to help debug:
1. Insert a `console.log()` in your edge function handler at the very beginning
2. Deploy edge function
3. View logs at https://app.supabase.com/project/_/functions/your-edge-function-name/logs and confirm that the logs shows up
4. If it does show up, try printing request header info, etc to try to sort out where the function is failing
5. If it doesn't show up, we can try to track down why your edge function isn't reachable

**Reply by Pasta** [reply]
so, you were 100% right in your first answer. my uri wasn't formatted correctly and i realized that when i added the log at the top of the function handler. it's working now =)

thanks so much!

**Comment by groyGetaway** [comment]
Hi @gregnr
Is the supabase client in the edge function supposed to contain the user?
`const {
    data: { user },
  } = await supabaseClient.auth.getUser();`  
This is returning null in the edge function.

On the front, I'm using fetchEventSource from @microsoft/fetch-event-source with `Bearer ${process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY}` as  the Authorization headers.


In the example :
```function createClient(req: Request) {
  return createSupabaseClient(
    // Supabase API URL - env var exported by default.
    Deno.env.get("SUPABASE_URL") ?? "",
    // Supabase API ANON KEY - env var exported by default.
    Deno.env.get("SUPABASE_ANON_KEY") ?? "",
    // Create client with Auth context of the user that called the function.
    // This way your row-level-security (RLS) policies are applied.
    {
      global: { headers: { Authorization: req.headers.get("Authorization")! } },
    }
  );
}
```

But so far, this has turned to be false (Auth context). Also I'm confused as to why it would work as I'm never sending the user token.

**Reply by gregnr** [reply]
Hey @groyGetaway! Your question is a bit of a tangent from the original discussion (consuming SSE stream client side). Could you please open a new discussion and we can debug there? Thanks!
